src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -1354,13 +1363,50 @@ public sealed override async Task<HttpResponseMessage> SendAsync(HttpRequestMess
             try
             {
                 // Send headers
-                http2Stream = await SendHeadersAsync(request, cancellationToken).ConfigureAwait(false);
+                bool shouldExpectContinue = request.Content != null && request.HasHeaders && request.Headers.ExpectContinue == true;
 
-                // Send request body, if any
-                await http2Stream.SendRequestBodyAsync(cancellationToken).ConfigureAwait(false);
+                http2Stream = await SendHeadersAsync(request, cancellationToken, flush: shouldExpectContinue).ConfigureAwait(false);
+                if (shouldExpectContinue)
+                {
+                    // Send header and wait a little bit to see if server sends 100, reject code or nothing.
+                    if (NetEventSource.IsEnabled) Trace($"Request content is not null, start processing 100-Continue.");
+                    await http2Stream.SendRequestBodyWithExpect100ContinueAsync(cancellationToken).ConfigureAwait(false);
+                }
+                else
+                {
+                    // Send request body, if any
+                    Task bodyTask = http2Stream.SendRequestBodyAsync(cancellationToken);
+                    // read response headers.
+                    Task responseHeadersTask = http2Stream.ReadResponseHeadersAsync();
+                    Task finished = await Task.WhenAny(new Task[] { bodyTask, responseHeadersTask}).ConfigureAwait(false);
+                    if (finished == bodyTask)
+                    {
+                        // Sending request body finished before getting headers.
+                        Task t = bodyTask;
+                        bodyTask = null;
+                        await t.ConfigureAwait(false);
```

Comment:
If this throws, what happens to potential failures in responseHeadersTask?

Reply:
We will call http2Stream.Dispose() bellow and that should make it fail. I'm not sure if you want to propagate the error or if you think we need to so something about it. Right now, I think we will simply throw first failure to caller.

Reply:
> I'm not sure if you want to propagate the error or if you think we need to so something about it.

If the body task fails, we don't need to also somehow propagate the failure in the responseHeadersTask, but that Task is almost certainly going to also fail with an exception, and if we don't do anything with it, that'll cause TaskScheduler.UnobservedTaskException to be raised.  It'd be better to observe the exception in a continuation and log it... that continuation can be hooked up only in the case in a catch around the await, such that we only pay for it if we're failing anyway.